cli/dump: dump all databases with a single command

C0rWin opened the pull request:
Fixes #28615.

This commit enables to dump all databases with a single command. New
flag is added to the `dump` command, namely `--all` which scans for all
available non-system databases and dump theirs content.

Signed-off-by: Artem Barger <bartem@il.ibm.com>

Release note (feature): enable to dump all databases with a single
command

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47387)
<!-- Reviewable:end -->

C0rWin commented:
Provides solution for #28615

dt commented on pkg/cli/dump_test.go:
```diff
@@ -490,3 +490,115 @@ ALTER TABLE orders VALIDATE CONSTRAINT fk_customer;
 	}
 }
+
+func TestDumpAllTables(t *testing.T) {
```
So postgres and thus we allow references across databases -- you can have an FK or a view in one DB refer to a table in another db, so we should be sure to include such a case here. I'm guessing that this means that above, you want to collect/accumulate all the tables for all the dbs _before_ you move on to actually dumping, and just emit all the create database statements together up front, then let the existing sort/emit/validate code run as-is, just over a larger accumulated set of tables.

C0rWin replied:
good point, thanks.

C0rWin replied:
updated code and added test case to reflect your suggestions

dt replied:
nice! can you add one more test case where we have _no_ user-created databases?


I think we also want to be dumping `defaultdb` since users can create tables in that one.

C0rWin replied:
I have intentionally omitted dumping of system databases, this is why I am using following query https://github.com/cockroachdb/cockroach/pull/47387/files#diff-57fbab5a71a2e66e38932592db7fbb9bR59

```
rows, err := conn.Query(`SELECT name FROM system.namespace WHERE id > $1 AND "parentID" = 0`, []driver.Value{minUserDescID})
```

C0rWin replied:
I guess, my question what would be the expected behavior? Shall `--dump-all` list and dump all databases including system? Add explicit flag to dump system databases?

dt replied:
i think system dbs in which users cannot create tables make sense to exclude, but `defaultdb` is special in that you _can_ create tables in it (i think?).

I'd say try to verify with a thought experiment / test-case: connect without a db specified (so you're using defaultdb), create a table, then dump with `--all`. I think a user could _expect_ the table they just created to be dumped. Thus I think we probably _should_ be dumping `defaultdb`. The tricky thing is that we likely can _not_ emit a CREATE for it (unless we add `IF NOT EXISTS`) because the importing cluster has its own `defaultdb`.

C0rWin replied:
> The tricky thing is that we likely can _not_ emit a CREATE for it (unless we add `IF NOT EXISTS`) because the importing cluster has its own `defaultdb`.

Now, when I think of that comments, it makes me wonder why `IF NOT EXISTS` is not appended now? I mean, for obvious reason `defaultdb` will exists (while honestly do not think it should be any different from normal databases), but there might be a collision with other names intentionally or not (normal names usually have very low entropy).

C0rWin replied:
> i think system dbs in which users cannot create tables make sense to exclude, but `defaultdb` is special in that you _can_ create tables in it (i think?).

totally make sense, but is there a way to distinguish dbs where user can create tables versus those where he cannot?

C0rWin replied:
also it makes me wonder how the query

```
SELECT name FROM system.namespace WHERE id > $1 AND "parentID" = 0
```

didn't hit `defaultdb`

C0rWin replied:
another thought, given we dump `defaultdb` would you expect to have 

```
CREATE DATABASE defaultdb;
```

in the resulted outcome? or since this is a default and potentially always existing database it should be 

```
USE defaultdb;
```

instead?

C0rWin replied:
> I'd say try to verify with a thought experiment / test-case: connect without a db specified (so you're using defaultdb), create a table, then dump with `--all`. I think a user could _expect_ the table they just created to be dumped.

I've came up with updated code + added a unit-test to cover the scenario with `defaultdb`.

C0rWin commented:
```
--- FAIL: TestReplicateQueueUpReplicate (47.02s)
    replicate_queue_test.go:218: condition failed to evaluate within 45s: replica count, want 3, current 1
        goroutine 1098670 [running]:
        runtime/debug.Stack(0xc002bf9e50, 0x4781240, 0xc00b7ae700)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
        github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x488b780, 0xc008484400, 0xc002bf9e50)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x6b
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicateQueueUpReplicate(0xc008484400)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replicate_queue_test.go:218 +0x5b8
        testing.tRunner(0xc008484400, 0x4091ef0)
        	/usr/local/go/src/testing/testing.go:909 +0xc9
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x350
```

seems like a flaky test, cause not directly related to changes.

knz reviewed (commented):
> Reviewed 2 of 5 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47387)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @C0rWin and @dt)

___
*[pkg/cli/dump.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47387#-M5LsF1-8IrtkYbRIdW6:-M5LsF10-TlNamVQOPCu:bpz6krb) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf2f318101182988df8baa04deb5a662838c01be/pkg/cli/dump.go#L105)):*
> ```Go
> 		}
> 		return fmt.Errorf("dump all while specifying database name at the same time is not supported, "+
> 			"these are mutual exclusive parameters, arguments provided are %s, and --dump-all flag is equal to %s", args, all)
> ```

nit: this conditional is too complex. Also the error message is too verbose. Try this:

```go
if dumpCtx.dumpAll && len(args) != 0 {
   return fmt.Errorf("cannot specify --%s and a specific database at the same time", cliflags.DumpAll.Name)
}
```

(also, a tip: printf with `%v` will render a `bool` as true/false, no need to make a string yourself)


<!-- Sent from Reviewable.io -->

C0rWin commented:
> ```go
> if dumpCtx.dumpAll && len(args) != 0 {
>    return fmt.Errorf("cannot specify --%s and a specific database at the same time", cliflags.DumpAll.Name)
> }
> ```

Thanks, pushed a fix per your suggestion.

dt commented:
This is looking great!

However, ugh, I apologize, but I just realized I missed a pretty big up-front design question which is how we want to qualify these names.

We currently match postgres in that we don't render these `create` statements with fully-qualified names (i.e. including the parent db name). We (and they) expect that we're only importing a dump of one DB or tables from one DB into exactly one DB (the one used by the importing session) and all the imported tables just get created in that DB. 

This is in fact intentional: we _want_ to be able to dump from one DB and import into a different one, and not qualifying the names allows that since we just resolve based on the session. However it means that as-is, dumping tables from more than one DB will end up flattening all of them into a single DB.

Obviously we'll need to lose that flexibility when dumping multiple DBs, but that seem understandable -- so i guess the next question is how.

One option is we could switch to emit fully-qualified names when we are dumping multiple databases. We'd need to switch ~everywhere though -- both in dump's rendering of INSERT statements as well as the server-rendered `CREATE` and `ALTER` statements in `crdb_internal.create_statements`, likely by adding another column that contains the qualified versions, and then dump could choose which one to read/emit based on single-vs-multi-DB mode.

Another option would be to emit `USE` statements in the dumpfile between blocks of CREATEs, INSERTS and ALTERS issued on behalf of each table. This though would then require we teach IMPORT to be aware of simulating a session search path and updating it as it parses.

One complicating factor to consider is that cockroach supports cross-DB references, which may or may not always need qualification, in single or multi mode.

I don't like either of these options much, but unfortunately since pg doesn't have a multi-db dump feature at all we don't have prior art to guide us, so I think we need to make our own call about what we want to do here. @knz might have some ideas.

C0rWin commented:
> Another option would be to emit `USE` statements in the dumpfile between blocks of CREATEs, INSERTS and ALTERS issued on behalf of each table. This though would then require we teach IMPORT to be aware of simulating a session search path and updating it as it parses.


I my humble opinion, the intended use case is to have more easier way to backup schemas. Therefore adding `USE` should be just fine and also less intrusive option in terms of code changes and focus area of this PR. 

However I'm fine with either direction as I can understand the pros of enhancing `IMPORT` to process `CREATE DATABASE`.

dt commented:
Yeah, that sounds good -- lets group the create statements by db and emit a `USE` statement before dumping each block, and then same for the inserts and trailing constraint additions.

I'm not sure why we have the topo sort here at all now that I think about it -- we have to move the constraint additions to the end _anyway_ because of cycles so I don't think we need it at all? I think we could just dump all the `CREATE`s a db at a time in alphabetical order, then same order for the inserts, then same for the contraints? I might be missing something though. Maybe @mjibson remembers?

mjibson commented:
Yes, that's my understanding too.

knz commented:
The topo sort was for view and sequence dependencies as much as foreign keys.

With views in particular ab alphabetical sort won't do.
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

knz commented:
Also I think interleaved tables need to be dumped from parent to child.

C0rWin commented:
> Also I think interleaved tables need to be dumped from parent to child.

That's the case now, isn't it?

knz commented:
Yes today because the code sorts the table in proper dump order it works.
It would not work properly if, as suggested by David above, we sorted them just in alphabetical order.

mjibson commented:
Is it needed for foreign keys? I thought those were all added after table creation and data insertion because there could be cyclic or self references? Ah, yes, I forgot about interleaved. If I'm correct about not needing topo sort for FKs, is interleaved the only reason it is needed? We should add a comment near the topo sort detailing exactly why it is needed, and also list the things it's not for. We've had this discussion many times now and we should document it somewhere.

knz commented:
>  is interleaved the only reason it is needed? 

Views come to mind as well as sequence definitions.

C0rWin commented:
> This is in fact intentional: we _want_ to be able to dump from one DB and import into a different one, and not qualifying the names allows that since we just resolve based on the session. However it means that as-is, dumping tables from more than one DB will end up flattening all of them into a single DB.

I think it was broken, since I realized that in the case of cross database foreign keys, it was/is dumped with db qualifier, e.g.

```
ALTER TABLE account ADD CONSTRAINT fk_person_id_ref_person FOREIGN KEY (person_id) REFERENCES dbb.public.person(id);
```

@dt @knz ?

C0rWin commented:
@dt updated PR per your suggestion.

C0rWin committed:
cli/dump: dump all databases with a single command

Fixes #28615.

This commit enables to dump all databases with a single command. New
flag is added to the dump command, namely --all which scans for all
available non-system databases and dump theirs content.

Signed-off-by: Artem Barger bartem@il.ibm.com

Release note (cli change): enable to dump all databases with a single
command.

dt commented:
Thanks for sticking with this through such a lengthy review/revision process and sorry again about forgetting to bring up the name qualifying earlier. 

And as a note to myself, we should also re-check this as part of an audit of cross-db interleave support, but that's an edge case that I don't think should hold up getting this merged now.

bors r=dt
